### ISSUE TITLE
How to make mu4e, mbsync, and keepassx-cli working together?

### ISSUE THREAD

**Issue by jpggithub** [djcb/mu#1973]
## Expected or desired behavior

I want to use my keepassxc database with keepassxc-cli to retrieve my mails from mu4e.
Using `mu4e-get-mail-command` and `mbsync`, mu4e should let me type the password of the database.

## Actual behavior

I can retrieve my mails from a terminal with  `mbsync johndoeAccountName`.
It asks me for the password, I give it and the mails are well synchronized.

In Emacs I use `(setq mu4e-get-mail-command "mbsync johndoeAccountName")`.

When I want to synchronise my mails from Emacs, a window displays 
`C: 0/1  B: 0/0  M: +0/0 *0/0 #0/0  S: +0/0 *0/0 #0/0` 
and a prompt asks for the password of the database but I can't type the password and the mails are not synchronized.

## Versions of mu, mu4e/emacs, operating system etc.
mu4e v1.4.15 (BTW it would be nice to have a mu4e-version function ;-) (something like emacs-version)
emacs 27.1
Debian testing

**Comment by djcb** [comment]
You can try playing with `mu4e~get-mail-password-regexp` etc. (see `mu4e-utils`), or you can integrate something like `zenity` in a shell script called from `mu4e-get-mail-command`.

**Comment by theophilusx** [comment]
Or another alternative is seeing if keypass can be configured to just
work directly with mbsync. I don't use keypass, but I have my imap
passwords stored in a gpg file and have the following in my .mbsyncrc

PassCmd "gpg -q --for-your-eyes-only --no-tty -d ~/.secure/gmail.gpg"

With this approach, when mu4e runs mbsync to synchronize my mail either
I'm prompted (via askpass) to enter the passphrase or if you have a
keyring manager (which I do on both Linux and Mac) and have authorized
the keyring manager to use my keys, I don't have to enter the passphrase
if I have already logged in) and I have the protection my passwords are
encrypted in gpg files. This also means I can ask mu4e to check for mail
every x minutes and I don't have to repeatedly enter the passphrase. 

So, if you can get keypass to work with mbsync, possibly using something
similar in principal to how I'm using gpg, you don't need to worry about
hacking things at the mu4e layer. As mbsync is also used by a number of
other mail clients, you may also have better luck finding how to use
keypass with mbsync than finding examples specific to mu4e. 

For smtp (sending) I just use .authinfo.gpg and all the same principals
work. 

-- 
Tim Cross

**Comment by thierryvolpiatto** [comment]
another alternative is to use authinfo mechanism provided by Emacs, copy the script below to /home/you and follow instructions in the usage section.

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Usage:
# Add:
# PassCmd "~/.mbsync.py -m machine_name -u you@your_provider.net -p 993"
# to mbsyncrc.
# Add a line in .authinfo.gpg file:
# machine machine_name port 993 login you@your_provider.net password xxxxxxxxx

from optparse import OptionParser
import re
import subprocess

parser = OptionParser()
parser.add_option("-m", "--machine", dest="machine", help ="The machine name as specified in auth source file")
parser.add_option("-u", "--user", dest="user", help ="User name (the email adress)")
parser.add_option("-p", "--port", dest="port", help ="Port")

(options, args) = parser.parse_args()
machine = options.machine
user = options.user
port = options.port


def get_authinfo_password(machine, login, port):
    # print machine
    # print user
    # print port
    s = "machine %s port %s login %s password ([^ ]*)\n" % (machine, port, login)
    (authinfo, erroutput) = subprocess.Popen("gpg -q --no-tty -d ~/.authinfo.gpg",
                                             shell=True,
                                             stdout=subprocess.PIPE,
                                             stderr=subprocess.STDOUT).communicate()
    return re.search(s,authinfo).group(1)
    
if __name__ == "__main__":
    print get_authinfo_password(machine, user, port)
```

**Comment by jpggithub** [comment]
Thanks a lot for your quick answers!

@djcb I try with `(setq mu4e~get-mail-password-regexp "^Enter password to unlock /home/use/keepassxc-database.kbdx: $")` and it works well. 
The problem is now that I have to give the password of the database each time I synchronize my mails :-/
I also test zenity but the same problem remains. I could store the password of the database on my computer with zenity and use it for the PassCmd of mbsync, but it seems to be very unsecure.

@theophilusx You are right, it will be more easier if mbsync and keepassxc could work better together but I didn't find a solution for that. In particular, keepassxc and keepassxc-cli don't speak together. So you can have your database open in keepassxc in one hand and must give your password if you use keepassxc-cli on the other hand...
I see many users that use gpg like you. I'm not at all against this solution but I find hard to master all this stuff.
I understand how private and public keys interact but I didn't understand why it is secure to have you private key is on the same computer that your .gpg files. 
Perhaps I missed something...

@thierryvolpiatto Thanks for you scrip. I think I have to master gpg first.

**Comment by theophilusx** [comment]
Actually, when you use pgp to encrypt a file in the means described
before, your really using symmetric encryption and not asymmetric
encryption (as is the case with private/public keys). All your really
doing is saying "Encrypt this file and use this passphrase as the key".
So unless someone knows your passphrase, even if they have access to
your computer, they cannot decrypt the file which has your password in
it. I suspect it is essentially the same basic algorithm as keypass uses
(I've never used keypass, but that is the basic model used by most
password management solutions). With pub/private key pairs, it is
critical to keep your private key private, but all private keys should
also have a strong passpharse so that even if they can gain access to
your private key, they still need the passpharse to use it.

The main risk in the model I use is that iff my login account is
compromised AND the individual is able to login to my account AND they
can login into my keychain, they may be able to access my gpg encrypted
files. However, if they have managed to already get that level of
access, it is game over anyway (especially as I have other layers of
protection, such as limiting access to only login at the physical
console (no ssh login etc) and on my laptop, full disk encryption, plus a few other
barriers).

The trick is to get the balance right. The level of protection needs to
be appropriate to the level of risk. The level of risk is determined by
likelihood and impact. There is also a play-off between convenience and
security.

For me, given what I do and how I use my system, the likelihood factor
is low to maybe moderate and impact is only low (I don't use email for
anything sensitive, don't store sensitive information in my email
mailboxes and generally treat email as an insecure communication
channel). So right now, the worst impact should my mail accounts be
compromised is inconvenience (and it would be extremely inconvenient).
I actually think the risk of unauthorised access to my account via my
provider is much higher than it being compromised via my
personal computer/device).

There have been times when the risks have been higher (specifically due
to corporate email for organisations I was working for which could
contain sensitive information, so calculating risk is an on-going
process which needs to be adjusted as circumstances change).

To a large extent, all modern computers come with password management
solutions 'built-in'. The real advantage of password managers like
keypass is in being able to have access to your vault of login
information across different devices and browsers. If you need that
facility, then a separate password management solution like keypass is
useful. If you don't, then you are likely adding additional complexity
which isn't providing any additional security and in fact may reduce it
(greater the complexity, greater the chance of something going wrong).

When it comes to mu4e, I also think it is important to recognise that
mu4e is somewhat agnostic regarding what the user uses to synchronise
their mailboxes. Using mbsync is just one tool which can be used with
mu4e. All mu4e really knows is that there is a command it can run to
update and sync between local mu4e mail folders and remote imap
services. How that tool obtains necessary credentials is really
something which needs to be resolved by that tool, not mu4e.

**Comment by jpggithub** [comment]
@theophilusx Thanks again for your detailed explanations. Yes, AFAIK `keepassxc` is very similar to `gpg` (used with symetric key).

I found other people that try to solve the problem encountered using `keepassxc` and `keepassxc-cli` with the same database (see https://github.com/keepassxreboot/keepassxc/issues/4513): it seems that the keepassxc model's lacks something similar to `gpg-agent`. Therefore the problem is not solved.

Finally, I configured `mu4e`, `mbsync`, and `msmtp` with `gpg` and `pass`. I keep `keepassxc` for browers.

It could be useful for people that want to do the same 'partial' migration from keepassxc to know that I used:
- https://marceloandrader.github.io/blog/2019/04/11/my-migration-from-keepassxc-to-pass-password-store/
- https://wiki.archlinux.org/index.php/GnuPG
- https://jherrlin.github.io/posts/emacs-mu4e/

I think this issue can be closed now because the problem comes from keepassxc, not mu4e.
